client/src/main/java/org/eclipse/hono/client/CommandConsumerFactory.java

```diff
@@ -36,6 +36,16 @@ static CommandConsumerFactory create(final HonoConnection connection) {
         return new CommandConsumerFactoryImpl(connection);
     }
 
+    /**
+     * Initializes the CommandConsumerFactory with the given GatewayMapper.
+     * <p>
+     * This needs to be called before calling any of the createCommandConsumer() methods.
+     *
+     * @param gatewayMapper The component mapping a command device id to the corresponding gateway device id.
+     * @throws NullPointerException if gatewayMapper is {@code null}.
+     */
+    void init(GatewayMapper gatewayMapper);
```

Comment:
why can't we pass in the gateway mapper to the factory method and the `CommandConsumerFactoryImpl` constructor? You can then override `AbstractHonoClientFactory`'s *connect()* and *disconnect()* method to start/sop both connections, or wouldn't that work?

Reply:
I thought about this, but decided against it.
This would mean that all the `ConnectionLifecycle` members of `AbstractProtocolAdapterBase` are connected in the `AbstractProtocolAdapterBase.connectToService` method - **except** the `gatewayMapper` field, where the connect invocation is hidden in the `CommandConsumerFactoryImpl` class.
This fact plus needing to override the `AbstractHonoClientFactory` methods makes this solution IMHO inferior to just adding the `init(GatewayMapper)` method.

Reply:
Or do you mean this along with removing the GatewayMapper field from the `AbstractProtocolAdapterBase` class?
I guess that would make exchanging the GatewayMapper implementation in a protocol adapter different from the way it would be done for the other factories.

Reply:
I guess we do not need (direct) access to the gateway mapper from the protocol adapters anyway, do we? Why not make it an implementation detail of the `CommandConsumerFactory`?

Reply:
I've explored this solution a bit more and think we can go with it. I've amended the commit with the corresponding changes.